### PR TITLE
docs(training): add DSpark speculative decoding design

### PR DESCRIPTION
**Pull request by khazic**
## What

Adds a design proposal (RFC) page, `docs/training/dspark-speculative-decoding.md`,
for DSpark speculative-decoding **draft training** in Megatron-Bridge. Tracks #5230.

## Why a design PR first

Megatron-Bridge has no draft-training subsystem today (MTP is the only
draft-adjacent path, and it is a Megatron-Core native module). DSpark's draft (a
semi-autoregressive parallel backbone plus Markov and confidence heads) is a new
neural module that must live either in Megatron-Core (like MTP) or as a
ModelOpt-injected module, and its parallel/loss integration can only be validated
on multi-GPU. So the draft module and its trainer belong in later, separately
reviewed PRs; this PR establishes the algorithm spec, the Bridge integration
shape, and the plan.

## Contents

- The DSpark draft: a semi-autoregressive block (parallel backbone attending to
  the frozen target's hidden states as context K/V) plus a Markov/RNN sequential
  head and a confidence head; the draft shares and freezes the target embedding
  and LM head.
- Training objective: position-weighted CE (0.1) + total-variation (0.9) +
  confidence BCE (1.0); acceptance = `1 - TV`, `tau = sum cumprod(accept) + 1`.
- Bridge mapping: follows the ModelOpt distillation subsystem pattern (provider
  hook, `forward_step_dspark`, a `post_training/dspark.py` loss, a thin
  `training/dspark.py` entry, and `dspark.*` HF weight mappings).

Reference implementation: NeMo AutoModel
`nemo_automodel/components/speculative/dspark/`. Paper:
[arXiv:2607.05147](https://arxiv.org/abs/2607.05147).

Docs-only; the page is registered in the "Training and Customization" toctree.

### CODE REVIEW

File: docs/training/dspark-speculative-decoding.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+Training measures acceptance analytically rather than by running rejection
+sampling. Per position, `accept_rate = clamp(1 - TV, 0, 1)`. The expected accepted
+prefix length of a block is `tau = sum_k cumprod(accept_rate)_k + 1` (a token
+survives only if every earlier token in its block is accepted; the `+1` counts the
+verified anchor token). `tau` is the training-time speedup proxy. The confidence
+head regresses to `accept_rate`; the paper applies a post-hoc Sequential
+Temperature Scaling calibration for serving.
+
+### Reference implementation (NeMo AutoModel)
```
Comment (yaoyu-33) [inline docs/training/dspark-speculative-decoding.md]: not needed...

**Review by yaoyu-33 — DISMISSED** [review]

File: docs/training/dspark-speculative-decoding.md (hunk 1 above)
Comment (khazic) [reply docs/training/dspark-speculative-decoding.md]: Removed that section. The pointer to the AutoModel reference in the overview stays.

File: docs/training/dspark-speculative-decoding.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+projection (`fc`) into the draft hidden size and used as the attention context.
+
+### Training objective
+
+The draft is trained with three position-weighted terms (weights `exp(-k/gamma)`,
+emphasizing earlier positions):
+
+- `L_ce` (weight 0.1): cross-entropy of the draft logits against the target's
+  teacher-forced next tokens.
+- `L_tv` (weight 0.9): total-variation distance `0.5 * ||p_draft - p_target||_1`
```
Comment (yaoyu-33) [inline docs/training/dspark-speculative-decoding.md]: **[High] This halves the dominant loss relative to the cited DSpark objective.** [Paper Eq. 10](https://arxiv.org/html/2607.05147#S3.E10) minimizes the weighted raw L1 distance `||p_draft - p_target||_1` with coefficient 0.9; the factor `1/2` appears in [Eq. 8](https://arxiv.org/html/2607.05147#S3.E8) only when forming the analytical acceptance/confidence target. The primary [DeepSpec implementation](https://github.com/deepseek-ai/DeepSpec/blob/005e03b81cec38b7da6399833d609ee89a2587f2/deepspec/modeling/dspark/loss.py#L73-L87) and current [NeMo AutoModel implementation](https://github.com/NVIDIA-NeMo/Automodel/blob/15da2681c6d827fa55c680e96dcaa8414db4cf88/nemo_automodel/components/speculative/dspark/loss.py#L59-L112) make the same distinction. With the documented coefficient, an implementation following this line gets half the intended training gradient (the proposed follow-up #5241 makes exactly that change). Please define the training term as weighted raw L1 and retain `0.5 * L1` only for the acceptance/confidence target, or explicitly document and validate a different coefficient contract.

File: docs/training/dspark-speculative-decoding.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+- `training/post_training/distillation.py`: the custom loss (`loss_func_kd`).
+- `training/gpt_step.py`: `forward_step_modelopt`, which selects that loss when the
+  model is a distillation model.
+- `training/distill.py`: a thin entry point, `distill(config) = pretrain(config,
+  forward_step_modelopt)`.
+
+Plus an `examples/distillation/` recipe and a `docs/training/` page. This
+provider-hook plus custom-forward-step plus custom-loss plus thin-entry pattern is
+the established, low-friction way to add a training subsystem here, and it is the
+recommended shape for DSpark.
```
Comment (yaoyu-33) [inline docs/training/dspark-speculative-decoding.md]: **[High] The recommended extension surface is already deprecated and conflates two different owners.** Current Bridge main warns that [`ModelProviderMixin`-based construction must migrate to `ModelConfig` + `ModelBuilder`](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/2086ee62cf2433c132b78778d43a99ea0ac1d2bf/src/megatron/bridge/models/model_provider.py#L241-L245), so a new dynamic provider subclass cannot be the promised stable API. The two choices above also lead to different contracts: an MCore-native design makes MCore own the parallel module/loss/sharded checkpoint, while a ModelOpt-injected design makes ModelOpt own the transformation/module/loss/export and Bridge only configures it. Current ModelOpt main already has an [HF DSpark plugin](https://github.com/NVIDIA/Model-Optimizer/blob/a21173a4d52cd7376a9e1ae274944d5983e3897c/modelopt/torch/speculative/plugins/hf_dspark.py) and [`DSparkExporter`](https://github.com/NVIDIA/Model-Optimizer/blob/a21173a4d52cd7376a9e1ae274944d5983e3897c/modelopt/torch/export/plugins/hf_spec_export.py#L510-L522), while Bridge currently pins [ModelOpt 0.46.0rc1](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/2086ee62cf2433c132b78778d43a99ea0ac1d2bf/pyproject.toml#L133-L137), which predates that support. Please choose one end-to-end owner, or give each alternative its own complete module/loss/construction/checkpoint/export contract, and make the Bridge hook builder-backed rather than prescribing the legacy provider path.

File: docs/training/dspark-speculative-decoding.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  token IDs), keeping the target and draft in one model.
+- Offline precomputation of target features, streamed to a draft-only training
+  job (as AutoModel's `precompute_dspark.py` does), which sidesteps holding the
+  target in memory but adds a cache format.
+
+### HF to/from Megatron conversion
+
+Because the draft reuses the target's embedding and LM head and adds `fc`,
+`markov_head`, `confidence_head`, and the draft decoder layers, the bridge mapping
+registry must gain `dspark.*` entries (QKV and gated-MLP splits for the draft
```
Comment (yaoyu-33) [inline docs/training/dspark-speculative-decoding.md]: **[High] `dspark.*` Bridge mappings are not implied until the frozen-weight and export schema are defined.** The references do not physically alias target parameters: AutoModel creates draft-owned `embed_tokens`/`lm_head`, [copies the target weights, and freezes the replicas](https://github.com/NVIDIA-NeMo/Automodel/blob/15da2681c6d827fa55c680e96dcaa8414db4cf88/nemo_automodel/components/speculative/dspark/draft_qwen3.py#L214-L230), then [controls their trainability separately](https://github.com/NVIDIA-NeMo/Automodel/blob/15da2681c6d827fa55c680e96dcaa8414db4cf88/nemo_automodel/components/speculative/dspark/draft_qwen3.py#L265-L282). That distinction determines PP placement, sharded save/resume, and whether serving export stores or reconstructs them. It also differs by owner: ModelOpt already owns its DSpark deployment schema/exporter, whereas an MCore-native checkpoint would need a separately specified HF schema before `AutoBridge` can map it. Please define alias-versus-replica storage, TP/vocab-sharded exact-L1 computation, selected/final-hidden PP transport, CP masking, online checkpoint versus offline-cache schema, and the repository that owns export; then list only the Bridge mappings that remain necessary.

**Review by yaoyu-33 — CHANGES_REQUESTED** [review]
Architecture disposition: changes requested. This is a useful docs-only RFC, but it does not yet define an implementable stable contract. Before merge, it should correct the objective scaling, choose one end-to-end owner (or specify two complete alternative contracts), and define the frozen-weight, distributed-loss, checkpoint, and export boundaries. No DSpark core implementation is requested in this PR.

**Comment by khazic** [comment]
Thanks @yaoyu-33, all three are addressed in 02156f95f. I verified each against the sources you cited before changing anything.

**Objective scaling.** You are right and the page was wrong. Eq. 10 minimizes the weighted raw L1 at coefficient 0.9, and the `1/2` belongs to the Eq. 8 acceptance label. AutoModel's `_compute_local_l1_term` accumulates `l1_dist_per_token * loss_weight_mask` with no `1/2`, and the `0.5` appears only in `_compute_accept_rate_3d`; DeepSpec splits it the same way. The page now defines `L_l1` as the raw weighted L1 and keeps `0.5 * L1` only in the acceptance label, and #5234 adds a regression asserting the pair together (disjoint distributions give `l1_loss == 2`, `accept_rate == 0`) so the two cannot drift again. I have closed #5241, which would have made exactly the change you flagged.

**Owner.** Settled on Megatron-Core native, with the reasoning written into a new "Decision" section rather than left implicit. The deciding fact is that ModelOpt's DSpark is HF-path only today: `plugins/` has `hf_dspark.py` and `modeling_dspark.py` and export has `DSparkExporter`, but the Megatron-side plugins are `megatron_eagle.py` and `megatron_medusa.py`, with no `megatron_dspark.py`. Going the ModelOpt route would mean landing a Megatron-native DSpark in Model-Optimizer and then waiting on the `nvidia-modelopt==0.46.0rc1` pin before any Bridge milestone could move. The consequence is spelled out too: Bridge owns the HF export schema, because `DSparkExporter` covers ModelOpt's own artifacts and not an MCore-native checkpoint.

**Deprecated extension surface.** Dropped. The construction path is now `ModelConfig` + `ModelBuilder` (a `DSparkModelConfig` naming a `DSparkModelBuilder`, mirroring `MambaModelConfig`/`MambaModelBuilder`), and the `dspark_provider.py` row is gone from the table along with the distillation-provider blueprint section that recommended it.

**Contracts that were missing.** The page now specifies replica-not-alias storage for the frozen embedding and LM head and why that choice drives PP placement, sharded save/resume, and export; shard-local exact L1 under TP (with gathered or top-k paths ruled out, since the acceptance label depends on its exactness); selected-hidden transport under PP; block-aligned anchor sampling under CP; the versioned offline feature-cache schema; and the checkpoint/export ownership split. It also documents the Megatron-Core loss return contract and its `calculate_per_token_loss=True` requirement, plus the measured memory contract for the probability L1. Both of those carry numbers from #5234's validation rather than estimates.

**Review by yaoyu-33 — APPROVED** [review]

**Comment by kamran-nvidia** [comment]
/ok to test [ac09592](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/5233/commits/ac0959283e21de5b05936903ea9bcc22e5a7e27d)
